File: apps/mobile/android/app/src/main/kotlin/com/example/mobile/ocr/SettleoraPaddleOcrEngine.kt
package com.example.mobile.ocr

import android.content.Context
import android.graphics.BitmapFactory
import com.paddle.ocr.EngineConfig
import com.paddle.ocr.PaddleOCRConfig
import com.paddle.ocr.engine.DetectionEngine
import com.paddle.ocr.engine.ORTSessionManager
import com.paddle.ocr.model.ModelConfig
import com.paddle.ocr.model.OCRError
import com.paddle.ocr.postprocess.BoxSorter
import com.paddle.ocr.postprocess.QuadTextCrop
import com.paddle.ocr.preprocess.RecPreprocessor
import com.paddle.ocr.util.BitmapUtils
import com.paddle.ocr.util.OpenCVUtils
import org.opencv.core.Core
import org.opencv.core.Mat

/**
 * Settleora's Android PaddleOCR/ONNX Runtime boundary.
 *
 * Detection runs once. The always-available common recognizer routes strong
 * common-script lines directly; only ambiguous lines receive a bounded pass
 * through the five Global Core specialists. A fixed calibrated selector then
 * chooses the compatible result without using phone locale.
 */
class SettleoraPaddleOcrEngine(context: Context) {
    private val appContext = context.applicationContext
    private val catalog = MobileOcrModelCatalog.load(appContext)
    private val config = settleoraPaddleOcrConfig()
    private val sessions = ORTSessionManager(appContext, EngineConfig())
    private val detector: DetectionEngine
    private val packs: List<RecognizerPack>

    val coldLoadTimeMs: Long
        get() = sessions.coldLoadTimeMs

    init {
        if (!OpenCVUtils.init(appContext)) {
            throw OCRError.ModelLoadFailed("opencv", IllegalStateException("OpenCV unavailable"))
        }

        val configuredPacks = catalog.recognizers.map { spec ->
            RecognizerPack(
                spec = spec,
                characters = ModelConfig.parse(appContext, spec.configAssetPath).characterList,
            )
        }
        try {
            sessions.loadModels(
                catalog.detection.modelAssetPath,
                configuredPacks.associate { it.spec.modelPackId to it.spec.modelAssetPath },
            )
        } catch (error: Throwable) {
            sessions.release()
            throw error
        }
        packs = configuredPacks
        detector = DetectionEngine(sessions, config)
    }

    fun recognize(imageBytes: ByteArray): SettleoraOcrRunResult {
        val sampleSize = validateInputBounds(imageBytes)
        val source = BitmapUtils.imdecodeBGR(imageBytes, sampleSize)
        if (source.empty()) {
            source.release()
            throw OCRError.InvalidImage()
        }
        if (!ReceiptOcrInputLimits.acceptsDimensions(source.cols(), source.rows())) {
            source.release()
            throw OCRError.ImageTooLarge()
        }

        val totalStart = System.currentTimeMillis()
        val sourceWidth = source.cols()
        val sourceHeight = source.rows()
        var sourceNeedsRelease = true
        val crops = mutableListOf<Mat>()
        return try {
            val detection = detector.detect(source)
            val sortedBoxes = BoxSorter.sortInReadingOrder(detection.boxes)
            if (!ReceiptOcrInputLimits.acceptsDetectedLineCount(sortedBoxes.size)) {
                throw OCRError.TooManyTextLines()
            }
            val blocks = mutableListOf<SettleoraOcrBlock>()
            var recognitionTimeMs = 0L

            val validBoxes = mutableListOf<Pair<Int, com.paddle.ocr.model.OCRBox>>()
            try {
                for ((order, box) in sortedBoxes.withIndex()) {
                    val fullResolutionCrop = QuadTextCrop.crop(source, box)
                    if (fullResolutionCrop.empty()) {
                        fullResolutionCrop.release()
                    } else {
                        val crop = try {
                            RecPreprocessor.resizeForRecognition(fullResolutionCrop)
                        } finally {
                            fullResolutionCrop.release()
                        }
                        validBoxes += order to box
                        crops += crop
                    }
                }
            } finally {
                source.release()
                sourceNeedsRelease = false
            }
            if (crops.isNotEmpty()) {
                val candidatesByLine = List(crops.size) { mutableListOf<ScriptCandidate>() }
                fun batchesFor(sourceCrops: List<Mat>, indices: Iterable<Int>) = indices
                    .groupBy { index -> recognitionBatchCapacity(sourceCrops[index]) }
                    .flatMap { (capacity, indices) ->
                        indices.sortedBy { index ->
                            sourceCrops[index].cols().toDouble() / sourceCrops[index].rows()
                        }
                            .chunked(capacity)
                    }

                fun recognizeBatch(
                    pack: RecognizerPack,
                    sourceCrops: List<Mat>,
                    targetCandidates: List<MutableList<ScriptCandidate>>,
                    batchIndices: List<Int>,
                ) {
                    val batchStart = System.currentTimeMillis()
                    val input = RecPreprocessor.preprocessBatch(
                        batchIndices.map { index -> sourceCrops[index] },
                    )
                    val decoded = sessions.runRecognitionDecoded(
                        pack.spec.modelPackId,
                        input.tensorData,
                        input.shape,
                        pack.characters,
                    )
                    recognitionTimeMs += System.currentTimeMillis() - batchStart
                    check(decoded.size == batchIndices.size) {
                        "Recognition batch size mismatch"
                    }
                    decoded.forEachIndexed { batchIndex, value ->
                        val lineIndex = batchIndices[batchIndex]
                        targetCandidates[lineIndex] += ScriptCandidate(
                            text = RecognizedTextNormalizer.normalize(
                                value.first.trim(),
                                pack.spec,
                            ),
                            confidence = value.second,
                            pack = pack.spec,
                        )
                    }
                }

                val commonPack = packs.singleOrNull { pack ->
                    ScriptEvidence.COMMON in pack.spec.acceptedScripts
                } ?: error("Exactly one common recognition pack is required")
                for (batchIndices in batchesFor(crops, crops.indices)) {
                    recognizeBatch(commonPack, crops, candidatesByLine, batchIndices)
                }

                val rotatedCrops = crops.map { crop ->
                    Mat().also { rotated -> Core.rotate(crop, rotated, Core.ROTATE_180) }
                }
                val rotatedCandidatesByLine = List(crops.size) {
                    mutableListOf<ScriptCandidate>()
                }
                try {
                    for (batchIndices in batchesFor(rotatedCrops, rotatedCrops.indices)) {
                        recognizeBatch(
                            commonPack,
                            rotatedCrops,
                            rotatedCandidatesByLine,
                            batchIndices,
                        )
                    }
                    // Orientation must not depend on the common recognizer for
                    // scripts that recognizer cannot read. Probe the bounded,
                    // catalog-pinned specialist set in both orientations before
                    // choosing the document direction.
                    for (pack in packs) {
                        if (pack === commonPack) continue
                        for (batchIndices in batchesFor(crops, crops.indices)) {
                            recognizeBatch(pack, crops, candidatesByLine, batchIndices)
                        }
                        for (batchIndices in batchesFor(rotatedCrops, rotatedCrops.indices)) {
                            recognizeBatch(
                                pack,
                                rotatedCrops,
                                rotatedCandidatesByLine,
                                batchIndices,
                            )
                        }
                    }
                    val uprightOrientationCandidates = candidatesByLine.map { candidates ->
                        ScriptRouteSelector.select(candidates) ?: candidates.first()
                    }
                    val rotatedOrientationCandidates = rotatedCandidatesByLine.map { candidates ->
                        ScriptRouteSelector.select(candidates) ?: candidates.first()
                    }
                    val reverseRecognition = ReceiptOrientationSelector.shouldRotate180(
                        uprightCandidates = uprightOrientationCandidates,
                        rotatedCandidates = rotatedOrientationCandidates,
                    )
                    if (reverseRecognition) {
                        crops.indices.forEach { index ->
                            crops[index].release()
                            crops[index] = rotatedCrops[index]
                            candidatesByLine[index].clear()
                            candidatesByLine[index] += rotatedCandidatesByLine[index]
                        }
                    }
                    val documentOrientation = ReceiptDocumentOrientation.select(
                        lineDimensions = validBoxes.map { (_, box) ->
                            val xs = box.points.map { it.x }
                            val ys = box.points.map { it.y }
                            (xs.maxOrNull()!! - xs.minOrNull()!!) to
                                (ys.maxOrNull()!! - ys.minOrNull()!!)
                        },
                        reverseRecognition = reverseRecognition,
                    )

                    candidatesByLine.forEachIndexed { index, candidates ->
                        val accepted = ScriptRouteSelector.select(candidates)
                        if (accepted != null && accepted.confidence >= config.recScoreThresh) {
                            val (order, box) = validBoxes[index]
                            blocks += SettleoraOcrBlock(
                                text = accepted.text,
                                confidence = accepted.confidence,
                                modelPackId = accepted.pack.modelPackId,
                                modelVersion = accepted.pack.modelVersion,
                                textDirection = ReceiptBlockOrder.textDirection(accepted.text),
                                order = order,
                                points = box.points.map { point ->
                                    documentOrientation.transform(
                                        SettleoraOcrPoint(point.x, point.y),
                                        sourceWidth,
                                        sourceHeight,
                                    )
                                },
                            )
                        }
                    }
                } finally {
                    rotatedCrops.forEachIndexed { index, rotated ->
                        if (rotated !== crops[index]) rotated.release()
                    }
                }
            }

            SettleoraOcrRunResult(
                blocks = ReceiptBlockOrder.normalize(blocks),
                detectionModelPackId = catalog.detection.modelPackId,
                detectionModelVersion = catalog.detection.modelVersion,
                runtime = RUNTIME_IDENTITY,
                coldLoadTimeMs = sessions.coldLoadTimeMs,
                detectionTimeMs = detection.timeMs,
                recognitionTimeMs = recognitionTimeMs,
                totalTimeMs = System.currentTimeMillis() - totalStart,
            )
        } finally {
            crops.forEach { it.release() }
            if (sourceNeedsRelease) source.release()
        }
    }

    fun release() = sessions.release()

    private fun validateInputBounds(imageBytes: ByteArray): Int {
        if (imageBytes.isEmpty()) throw OCRError.InvalidImage()
        if (!ReceiptOcrInputLimits.acceptsEncodedSize(imageBytes.size)) {
            throw OCRError.ImageTooLarge()
        }

        val bounds = BitmapFactory.Options().apply { inJustDecodeBounds = true }
        BitmapFactory.decodeByteArray(imageBytes, 0, imageBytes.size, bounds)
        val width = bounds.outWidth
        val height = bounds.outHeight
        if (width <= 0 || height <= 0) throw OCRError.InvalidImage()
        val sampleSize = ReceiptOcrInputLimits.sampleSizeFor(width, height)
        if (sampleSize == null) {
            throw OCRError.ImageTooLarge()
        }
        return sampleSize
    }

    private fun recognitionBatchCapacity(crop: Mat): Int {
        val normalizedWidth = kotlin.math.ceil(48.0 * crop.cols() / crop.rows())
            .toInt()
            .coerceAtMost(3200)
        return when {
            normalizedWidth > 1280 -> 1
            normalizedWidth > 640 -> minOf(2, config.recBatchSize)
            else -> config.recBatchSize
        }
    }

    private data class RecognizerPack(
        val spec: RecognizerSpec,
        val characters: List<String>,
    )

    companion object {
        const val CHANNEL_NAME = "com.settleora.mobile/receipt_ocr"
        const val METHOD_RECOGNIZE = "recognize"
        const val RUNTIME_IDENTITY = "onnxruntime-android:1.21.1:cpu"

    }
}

internal fun settleoraPaddleOcrConfig() = PaddleOCRConfig(
    detMaxSideLimit = 1600,
    detThresh = 0.2f,
    detBoxThresh = 0.45f,
    detUnclipRatio = 1.4f,
    detMaxCandidates = 3000,
    recScoreThresh = 0.1f,
    recBatchSize = 4,
)

data class SettleoraOcrRunResult(
    val blocks: List<SettleoraOcrBlock>,
    val detectionModelPackId: String,
    val detectionModelVersion: String,
    val runtime: String,
    val coldLoadTimeMs: Long,
    val detectionTimeMs: Long,
    val recognitionTimeMs: Long,
    val totalTimeMs: Long,
) {
    fun toChannelValue(): Map<String, Any> = mapOf(
        "blocks" to blocks.map { it.toChannelValue() },
        "detectionModelPackId" to detectionModelPackId,
        "detectionModelVersion" to detectionModelVersion,
        "runtime" to runtime,
        "coldLoadTimeMs" to coldLoadTimeMs,
        "detectionTimeMs" to detectionTimeMs,
        "recognitionTimeMs" to recognitionTimeMs,
        "totalTimeMs" to totalTimeMs,
    )
}

data class SettleoraOcrBlock(
    val text: String,
    val confidence: Float,
    val modelPackId: String,
    val modelVersion: String,
    val textDirection: String,
    val order: Int,
    val row: Int = 0,
    val points: List<SettleoraOcrPoint>,
) {
    fun toChannelValue(): Map<String, Any> = mapOf(
        "text" to text,
        "confidence" to confidence.toDouble(),
        "modelPackId" to modelPackId,
        "modelVersion" to modelVersion,
        "textDirection" to textDirection,
        "order" to order,
        "row" to row,
        "points" to points.map { it.toChannelValue() },
    )
}

data class SettleoraOcrPoint(val x: Float, val y: Float) {
    fun toChannelValue(): Map<String, Double> = mapOf(
        "x" to x.toDouble(),
        "y" to y.toDouble(),
    )
}

File: apps/mobile/ios/Runner/PaddleOCR/SettleoraPaddleOcrEngine.swift
import CoreGraphics
import Foundation
import ImageIO

struct SettleoraOcrPoint {
  let x: Double
  let y: Double
  var channelValue: [String: Double] { ["x": x, "y": y] }
}

struct SettleoraOcrBlock {
  let text: String
  let confidence: Float
  let modelPackID: String
  let modelVersion: String
  let textDirection: String
  var order: Int
  var row: Int
  let points: [SettleoraOcrPoint]

  var channelValue: [String: Any] {
    [
      "text": text,
      "confidence": Double(confidence),
      "modelPackId": modelPackID,
      "modelVersion": modelVersion,
      "textDirection": textDirection,
      "order": order,
      "row": row,
      "points": points.map(\.channelValue),
    ]
  }
}

struct SettleoraOcrRunResult {
  let blocks: [SettleoraOcrBlock]
  let detectionModelPackID: String
  let detectionModelVersion: String
  let coldLoadTimeMs: Int
  let detectionTimeMs: Int
  let recognitionTimeMs: Int
  let totalTimeMs: Int

  var channelValue: [String: Any] {
    [
      "blocks": blocks.map(\.channelValue),
      "detectionModelPackId": detectionModelPackID,
      "detectionModelVersion": detectionModelVersion,
      "runtime": "onnxruntime-objc:1.24.3:cpu",
      "coldLoadTimeMs": coldLoadTimeMs,
      "detectionTimeMs": detectionTimeMs,
      "recognitionTimeMs": recognitionTimeMs,
      "totalTimeMs": totalTimeMs,
    ]
  }
}

final class SettleoraPaddleOcrEngine {
  private struct Recognizer {
    let spec: RecognizerSpec
    let preprocessor: RecPreprocessor
    let decoder: CTCDecoder
  }

  private let catalog: MobileOcrModelCatalog
  private let sessions: ORTSessionManager
  private let detector: DetectionEngine
  private let recognizers: [Recognizer]
  private let coldLoadTimeMs: Int

  static func make() async throws -> SettleoraPaddleOcrEngine {
    let start = CFAbsoluteTimeGetCurrent()
    let catalog = try MobileOcrModelCatalog.load()
    let sessions = ORTSessionManager()
    try await sessions.loadModels(
      detectionPath: catalog.detection.modelPath,
      recognizerPaths: Dictionary(uniqueKeysWithValues: catalog.recognizers.map {
        ($0.modelPackID, $0.modelPath)
      })
    )
    let detector = try DetectionEngine(
      sessionManager: sessions,
      configPath: catalog.detection.configPath
    )
    let recognizers = try catalog.recognizers.map { spec in
      let config = try InferenceConfig.load(from: spec.configPath)
      return Recognizer(
        spec: spec,
        preprocessor: try RecPreprocessor(config: config),
        decoder: try CTCDecoder(config: config)
      )
    }
    return SettleoraPaddleOcrEngine(
      catalog: catalog,
      sessions: sessions,
      detector: detector,
      recognizers: recognizers,
      coldLoadTimeMs: milliseconds(since: start)
    )
  }

  private init(
    catalog: MobileOcrModelCatalog,
    sessions: ORTSessionManager,
    detector: DetectionEngine,
    recognizers: [Recognizer],
    coldLoadTimeMs: Int
  ) {
    self.catalog = catalog
    self.sessions = sessions
    self.detector = detector
    self.recognizers = recognizers
    self.coldLoadTimeMs = coldLoadTimeMs
  }

  func recognize(_ imageData: Data) async throws -> SettleoraOcrRunResult {
    guard !imageData.isEmpty, imageData.count <= 25 * 1024 * 1024 else {
      throw SettleoraOcrError.imageTooLarge
    }
    let image = try EncodedImageCodec.cgImage(fromEncodedData: imageData)
    guard image.width <= 8192, image.height <= 8192,
          image.width * image.height <= 16_000_000 else {
      throw SettleoraOcrError.imageTooLarge
    }

    let totalStart = CFAbsoluteTimeGetCurrent()
    let runtime = OCRRuntimeParams(
      textDetLimitSideLen: 1600,
      textDetLimitType: "max",
      textDetMaxSideLimit: 1600,
      textDetThresh: 0.2,
      textDetBoxThresh: 0.45,
      textDetUnclipRatio: 1.4,
      textRecBatchSize: 4,
      textRecScoreThresh: 0.1
    )
    let detection = try await detector.detect(image, runtimeParams: runtime)
    let boxes = BoxSorter.sortInReadingOrder(detection.boxes)
    guard boxes.count <= 128 else { throw SettleoraOcrError.tooManyLines }
    let crops = try boxes.map { try QuadTextCrop.crop(image, polygon: $0.points) }

    guard let common = recognizers.first(where: { $0.spec.acceptedScripts.contains(.common) }) else {
      throw SettleoraOcrError.invalidCatalog
    }
    let recognitionStart = CFAbsoluteTimeGetCurrent()
    var candidates = try await recognize(common, crops: crops).map { [$0] }
    let rotatedCrops = try crops.map(rotate180)
    let rotatedCommon = try await recognize(common, crops: rotatedCrops)
    let rotateDocument = ReceiptOrientationSelector.shouldRotate180(
      upright: candidates.compactMap(\.first),
      rotated: rotatedCommon
    )
    let selectedCrops = rotateDocument ? rotatedCrops : crops
    if rotateDocument {
      for index in candidates.indices { candidates[index] = [rotatedCommon[index]] }
    }

    // Correctness-first bounded routing: every verified specialist sees every
    // detected crop, so a common-model hallucination cannot suppress Arabic,
    // Thai, or another mandatory Global Core script.
    for recognizer in recognizers where recognizer.spec != common.spec {
      let results = try await recognize(recognizer, crops: selectedCrops)
      for index in results.indices { candidates[index].append(results[index]) }
    }

    var blocks: [SettleoraOcrBlock] = []
    for index in candidates.indices {
      guard let accepted = ScriptRouteSelector.select(candidates[index]),
            accepted.confidence >= 0.1 else { continue }
      let points = boxes[index].points.map { point -> SettleoraOcrPoint in
        if rotateDocument {
          return SettleoraOcrPoint(
            x: Double(max(0, min(image.width - 1, image.width - 1 - Int(point[0])))),
            y: Double(max(0, min(image.height - 1, image.height - 1 - Int(point[1]))))
          )
        }
        return SettleoraOcrPoint(x: Double(point[0]), y: Double(point[1]))
      }
      blocks.append(SettleoraOcrBlock(
        text: accepted.text,
        confidence: accepted.confidence,
        modelPackID: accepted.pack.modelPackID,
        modelVersion: accepted.pack.modelVersion,
        textDirection: ReceiptBlockOrder.textDirection(accepted.text),
        order: index,
        row: 0,
        points: points
      ))
    }
    let recognitionTimeMs = milliseconds(since: recognitionStart)
    return SettleoraOcrRunResult(
      blocks: ReceiptBlockOrder.normalize(blocks),
      detectionModelPackID: catalog.detection.modelPackID,
      detectionModelVersion: catalog.detection.modelVersion,
      coldLoadTimeMs: coldLoadTimeMs,
      detectionTimeMs: Int((detection.totalTime * 1000).rounded()),
      recognitionTimeMs: recognitionTimeMs,
      totalTimeMs: milliseconds(since: totalStart)
    )
  }

  private func recognize(_ recognizer: Recognizer, crops: [CGImage]) async throws -> [ScriptCandidate] {
    var output: [ScriptCandidate] = []
    for batch in crops.chunked(maxCount: 4) {
      let input = try recognizer.preprocessor.preprocessBatch(batch)
      let raw = try await sessions.runRecognition(
        packID: recognizer.spec.modelPackID,
        inputData: input.tensorData,
        shape: input.tensorShape
      )
      let decoded = try recognizer.decoder.decodeBatch(outputData: raw.data, outputShape: raw.shape)
      output.append(contentsOf: decoded.map {
        ScriptCandidate(
          text: RecognizedTextNormalizer.normalize($0.text.trimmingCharacters(in: .whitespacesAndNewlines), pack: recognizer.spec),
          confidence: $0.confidence,
          pack: recognizer.spec
        )
      })
    }
    return output
  }

  private func rotate180(_ image: CGImage) throws -> CGImage {
    guard let colorSpace = image.colorSpace ?? CGColorSpace(name: CGColorSpace.sRGB),
          let context = CGContext(
            data: nil,
            width: image.width,
            height: image.height,
            bitsPerComponent: 8,
            bytesPerRow: image.width * 4,
            space: colorSpace,
            bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue
          ) else { throw SettleoraOcrError.invalidImage }
    context.translateBy(x: CGFloat(image.width), y: CGFloat(image.height))
    context.rotate(by: .pi)
    context.draw(image, in: CGRect(x: 0, y: 0, width: image.width, height: image.height))
    guard let result = context.makeImage() else { throw SettleoraOcrError.invalidImage }
    return result
  }
}

private func milliseconds(since start: CFAbsoluteTime) -> Int {
  Int(((CFAbsoluteTimeGetCurrent() - start) * 1000).rounded())
}

private extension Array {
  func chunked(maxCount: Int) -> [[Element]] {
    guard !isEmpty else { return [] }
    return stride(from: 0, to: count, by: maxCount).map {
      Array(self[$0..<Swift.min($0 + maxCount, count)])
    }
  }
}
